[gpio] Refactor SAM signals API

henrikssn opened the pull request:
Goals:
- Compilaton errors are understandable (contain necessary information about the context that lead to the error).
- Completely in C++ templates, allows peripherals to perform further validation.
- Allows the user to select the intended peripheral signal (e.g. TX, SCK, MOSI, SDA)
- Allows peripherals to take arguments in arbitrary order (my gift to @salkinium )
- Deduces the pin function at compile time
- Deduces the Peripheral pad at compile time (e.g. Sercom Pad)
- As close to the current signals API as possible

Fixes #481.

salkinium commented:
This looks great on first viewing! The STM32 API was still C++11 based, which didn't have variadic templates, so we generated everything with Python. I've always been busy fixing lbuild/modm-devices/tooling stuff that I don't have enough time to properly engage on the C++17 API front. There's a lot more to improve.

I'm still busy with TinyUSB and then I'd like to review/merge your OTA code, so I'll check this out, but probably not fast.

PS: Please fix your formatting style, or I'm going to have a hard time reviewing this.

henrikssn commented:
> This looks is great on first viewing! The STM32 API was still C++11 based, which didn't have variadic templates, so we generated everything with Python. I've always been busy fixing lbuild/modm-devices/tooling stuff that I don't have enough time to properly engage on the C++17 API front. There's a lot more to improve.
> 
> I'm still busy with TinyUSB and then I'd like to review/merge your OTA code, so I'll check this out, but probably not fast.

Yeah I learned a lot about C++17 while doing this, which was my main motivation. It believe it should be fairly straight-forward to make the STM32 API follow the pattern, but I'd like to make that a separate PR.

The OTA code is still WIP and not ready to merge, I would recommend you start here and then look at the fiber PR.

> PS: Please fix your formatting style, or I'm going to have a hard time reviewing this.

Fixed, I also added the .clang-format file which I used so that I can stay more consistent in the future.

PS: How I dislike those hard tabs...

salkinium commented on src/modm/board/feather_m0/board.hpp:
```diff
@@ -100,18 +97,18 @@ struct SystemClock
-using LoggerDevice = modm::IODeviceWrapper< Uart0, modm::IOBuffer::BlockIfFull >;
-using Leds = SoftwareGpioPort< Led >;
+using LoggerDevice = modm::IODeviceWrapper<Uart0, modm::IOBuffer::BlockIfFull>;
+using Leds = Led;
```
This does not have the same semantics. `Leds::write(uint32_t)` accepts a bit vector, `Led::write(bool)` only accepts a single bit.

henrikssn replied:
Yes, but a bit vector is implicitly convertible to a bool, so it should be ok in all cases.

Here is the line I am trying to make it work with: https://github.com/modm-io/modm/blob/3072005ddf57cad660a095335573a7614bf533af/src/modm/board/board.cpp.in#L50

Is it expected that it only will blink the first led?

rleh replied:
What is the benefit of removing the `SoftwareGpioPort<>` here?
I would like to keep it in order to have the same `board.hpp` files for further automation and testing in the future (#359).

henrikssn replied:
It wasn't implemented on SAM properly and the new `Gpio<>` is drop-in compatible with a `SoftwareGpioPort<>` that only contains one port (which is the case for all SAM boards at the moment).

Maybe we can move the `SoftwareGpioPort<>` to a common library after stm32/avr supports the same signals API?

salkinium commented on src/modm/board/feather_m0/board.hpp:
```diff
@@ -100,18 +97,18 @@ struct SystemClock
 	SystemClock::enable();
 	SysTickTimer::initialize<SystemClock>();
-	Uart0::connect<Rx::Pad3, Tx::Pad2>();
+	Uart0::connect<Rx, Tx>();
```
I'd prefer to have Pin and Signal name in the same line.
```suggestion
	Uart0::connect<Rx::Rx, Tx::Tx>();
```

henrikssn replied:
But they are totally different, for example you can use the `Rx` signal from the `Tx` pin (`Tx::Rx`) if we make that decision, which makes no sense to me.

The problem is that the signal to pin mapping is much more flexible on SAM, so we need to be a bit careful keeping them apart.

salkinium commented on src/modm/board/feather_m0/board.hpp:
```diff
@@ -13,57 +13,55 @@
-using Rx	= GpioA11;
-using Tx	= GpioA10;
+using Rx = Gpio<A11>::Rx;
+using Tx = Gpio<A10>::Tx;
```
Hm?

salkinium commented on src/modm/platform/gpio/sam/config.hpp.in:
```diff
@@ -0,0 +1,108 @@
+
+enum class PortName {
+  A,
+  B,
```
Should be generated?

henrikssn replied:
Yes, although I couldn't think of a SAM device with only one port (and afaik there is none with more than 2)

salkinium commented on src/modm/platform/gpio/sam/pin.hpp:
```diff
@@ -0,0 +1,302 @@
+{
+	Rx,
+	Tx,
+	ExtInt,
```
Shouldn't there be more?

henrikssn replied:
We don't have any others implemented for SAM atm. I'll work on extending this list next.

salkinium reviewed (commented):
I like this very, very much! It's so much cleaner and much faster to generate/compile than the STM32 solution, which generates 200+ files just for the GPIOs, which is just dumb.

I added a closer-to-modm-style clang format, but I guess we're going to have to use `// clang-format off` for some of the short function stuff. You still have to set hardtabs in your editor for the template stuff. Theres a `.editorconfig` for that purpose, it's supposed to be editor independent.

I would still like to see an alias for `using GpioA00 = Gpio<A00>;` for backwards compatiblity? Or perhaps just for STM32? I'm not sure I like having a `A00` type floating around globally (due to `using namespace modm::platform;`), so perhaps it should be `using GpioA00 = Gpio<GpioConfigA00>;`?

henrikssn reviewed (commented):
> I like this very, very much! It's so much cleaner and much faster to generate/compile than the STM32 solution, which generates 200+ files just for the GPIOs, which is just dumb.

Thanks for the review!
> 
> I added a closer-to-modm-style clang format, but I guess we're going to have to use `// clang-format off` for some of the short function stuff. You still have to set hardtabs in your editor for the template stuff. Theres a `.editorconfig` for that purpose, it's supposed to be editor independent.

I disagree your clang-format is substantially closer to modm-style than the one I added. I force-pushed the branch back to where it originally was, in the future, please make changes in commits so that we can review the differences. I spent multiple hours aligning it with the modm style so it would be good to compare our efforts...

Also feel free to comment on the style in this PR, then we can see how we can fix that in the clang-format file.

I used this command to assess quality of the clang-format file: `find src/modm | grep -e "\.\(c\|h\|hpp\|cpp\)\$" | xargs clang-format -n 2>&1 | wc -l` (lower is better)

Your config: 143759
My config: 143800

Note that my config does allow the short functions, so no need for the `// clang-format off`.

> 
> I would still like to see an alias for `using GpioA00 = Gpio<A00>;` for backwards compatiblity? Or perhaps just for STM32? 

I'm fine with adding the using declarations, but it should be a separate templated header file (which we could probably share between the platforms).

> I'm not sure I like having a `A00` type floating around globally (due to `using namespace modm::platform;`), so perhaps it should be `using GpioA00 = Gpio<GpioConfigA00>;`

How about `PinA01` then? `GpioConfigA00` is too long for my taste.

salkinium commented on src/modm/board/feather_m0/board.hpp:
```diff
@@ -13,57 +13,55 @@
 /// samd21g18a running at 48MHz generated from the external 32.768 KHz crystal
-struct SystemClock
-{
+struct SystemClock {
```
This brace should be in the next line.

henrikssn replied:
Nice catch, done (in .clang-format)

salkinium commented on src/modm/platform/gpio/sam/config.hpp.in:
```diff
@@ -0,0 +1,108 @@
+
+%% endfor
+
+} // namespace modm::platform
```
This file needs to be converted to tabs and the { put on a newline.

henrikssn replied:
Braces done, I still need to figure out how to configure my editor (VSCode) for properly emitting tabs

salkinium replied:
I've added a `.editorconfig` to make GitHub render tabs as 4 spaces. Perhaps you can use the same mechanism. Something like this perhaps:
https://marketplace.visualstudio.com/items?itemName=EditorConfig.EditorConfig

henrikssn replied:
I tried it but it makes it impossible to edit the python files, as there we use spaces and python considers the difference important

salkinium replied:
Then we need to add a filter for python files to the config.

salkinium replied:
Try adding this to `.editorconfig`, that should make the tabs go away for the Python files.
```
[*.{lb,py}]
indent_style = space
indent_size = 4
```

salkinium commented on examples/samd/interrupt/main.cpp:
```diff
@@ -16,26 +16,27 @@ using namespace std::chrono_literals;
+	ExternalInterrupt::initialize();
+	ExtInt<3>::initialize(&isr);
+	ExtInt<3>::connect<D12>();
+	while (1) {
```
{ on new line.

henrikssn replied:
Done (also in .clang-format)

salkinium commented:
I'm currently super busy with manufacturing 100 railway signals, so I'm AFK for a few weeks.
If you can fix the { on the newline, then I'm happy to merge this.

salkinium reviewed (approved):
Need to make the CI pass and rebase and squash onto develop please.

salkinium commented on src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp:
```diff
@@ -4,6 +4,7 @@
  * Copyright (c) 2012-2017, Niklas Hauser
  * Copyright (c) 2014, Sascha Schade
  * Copyright (c) 2018, Raphael Lehmann
+ * Copyright (c) 2020, Erik Henriksson
```
And what did you change? This giant diff really isn't selling me on using clang format and basically just demotivates me to maintain modm entirely. Like, how do you expect me to review this?

salkinium replied:
Much, much better now. Thank you!

salkinium commented on src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp:
```diff
@@ -48,20 +45,23 @@ modm::platform::BitBangSpiMaster<Sck, Mosi, Miso>::connect()
+					   (Connector::template IsValid<Miso> and
+						sizeof...(Signals) == 3 and ContainsMiso)),
+				  "BitBangSpiMaster<Sck, Mosi, Miso> can only connect to the "
+				  "same Sck, Mosi and Miso signals of the declaration!");
```
How in hell is this more readable?

henrikssn replied:
I believe this is caused by indenting the line wrap 4 tabs (should only be one). Let me fix that...

salkinium commented on src/modm/platform/i2c/bitbang/bitbang_i2c_master_impl.hpp:
```diff
@@ -111,19 +129,25 @@ modm::platform::BitBangI2cMaster<Scl, Sda>::start(modm::I2cTransaction *transact
+				DEBUG_SW_I2C(
+					(starting.next == modm::I2c::OperationAfterStart::Read)
+						? 'r'
+						: 'w');
```
Srsly? No.

henrikssn replied:
Done

salkinium commented:
Perhaps using clang format is ok on new files, but we're not reformatting entire swathes of modm, just because one line changes. I'm not wasting my free time fighting an autoformatter.

salkinium commented:
I'm sorry I'm exploding at you, but these are exactly the same issues that I first encountered ~3 years ago when investigating using clang format for modm (during the porting from xpcc) and it's very frustating to see all the failing corner cases. I like the idea, but I'm going to have to deal with the fallout of this decision for many, many years going forward and I don't want to burn out on this project from preventable causes.

salkinium commented:
And why did the STM32 CMSIS header and FreeRTOS submodules change?

salkinium commented:
\*challening the inner grumpy old maintainer\* get off my lawn!1!!

henrikssn commented:
Rebased + fixed CI

henrikssn commented:
> I'm sorry I'm exploding at you, but these are exactly the same issues that I first encountered ~3 years ago when investigating using clang format for modm (during the porting from xpcc) and it's very frustating to see all the failing corner cases. I like the idea, but I'm going to have to deal with the fallout of this decision for many, many years going forward and I don't want to burn out on this project from preventable causes.

If I look at the changes, there are more cases where the existing style is inconsistent with the modm style than there are cases where clang-format IMO is wrong.

For example, the contribution guide specifies no space after template keyword, yet that is the case in all of the bitbang implementations. Also, brace wrapping of if/else conditions is done very inconsistently (and this one isn't well defined in the style guide). At some point, you need to bite the apple and decide which one is actually correct.

Obviously, if the goal is to have inconsistent style everywhere, then we should not use clang format _[grumpy new contributor mode on] but then I expect no more comments on the style of my commits... [/grumpy]_

salkinium commented:
> if the goal is to have inconsistent style everywhere

The goals are in order:

1. Don't burn out the maintainer.
2. Have somewhat structured, reviewable PRs, so we don't burn out the maintainer.
3. Have somewhat consistent, readable code, so we don't burn out the maintainer.
4. Don't burn out the maintainer.
 
All are pretty subjective, but forcing me to *search* the diff in a reformatted file is not acceptable. I also contribute to other projects and I don't reformat entirely unrelated lines of code because I want my changes merged and not bikeshed about code formatting.

Our formatting guide is basically about quickly grokking code, by visually matching vertically scoped code elements, like most control structured and function declarations.

1. Use tabs, yes we hate it too, but it's too late to change.
2. Put braces in a new line unless it becomes an unreadable mess.
3. Put the function name on a newline unless it becomes an unreadable mess.
4. Limit yourself to <100 chars unless it becomes an unreadable mess.

This is a cooperation between you and me, not a negotiation. You adapt your style a little and I look the other way and then we can both be maximize our happiness. This has worked quite well for most contributors and it can work well for us too.

> the contribution guide specifies no space after template keyword, yet that is the case in all of the bitbang implementations.

The bit-bang implementation is adapted from code from 2006 I believe. Much modm code is over a decade old. When we ported xpcc to modm, we experimented with clang-format to change the tabs to spaces, but it was just too many changes it would have double the repository size and made it impossible to use git blame for anything anymore.

Which is why using clang-format on purely added files is very acceptable to me, but let's not plough over the entire repo on principle.

henrikssn commented:
I reverted the style changes to the bitbang* files and rebased/squashed against develop.

salkinium reviewed (approved):
Excellent.

rleh commented on .editorconfig:
```diff
@@ -11,3 +11,7 @@ insert_final_newline = true
 trim_trailing_whitespace = false
+
+[*.{lb,py}]
+indent_style = space
```
```suggestion
indent_style = tab
```

rleh replied:
> `[*.{lb,py}]`

Nevermind...

rleh commented on src/modm/board/feather_m0/board.hpp:
```diff
@@ -88,8 +87,7 @@ struct SystemClock
 
-	static bool inline
-	enable()
+	static bool inline enable()
```
Looks unintentional to me.

henrikssn replied:
Fixed

rleh commented on src/modm/platform/gpio/sam/module.md:
```diff
@@ -0,0 +1,76 @@
+As a few examples, the following situations will crash compilation:
+
+- Usage of drivers which don't exist on the pin
+- Passing a pin which has a connection to  the driver but not the correct instance
```
```suggestion
- Passing a pin which has a connection to the driver but not the correct instance
```

henrikssn replied:
Done

rleh commented on src/modm/platform/gpio/sam/unused.hpp:
```diff
@@ -60,15 +61,6 @@ class GpioUnused : public Gpio, public ::modm::GpioIO
 	// end documentation inherited
 	static void configure(InputType) {}
 	static void setInput(InputType) {}
-	// External Interrupts
-	static void enableExternalInterrupt() {}
-	static void disableExternalInterrupt() {}
-	static void enableExternalInterruptVector(const uint32_t) {}
-	static void disableExternalInterruptVector() {}
-	static void setInputTrigger(const InputTrigger) {}
-	static bool getExternalInterruptFlag() { return false; }
-	/// Reset the interrupt flag in the interrupt routine.
-	static void acknowledgeExternalInterruptFlag() {}
```
Why are all the functions removed from `GpioUnused`?

salkinium replied:
The SAM port has its own separate peripheral for external interrupts, thus this API is not part of the (platform-specific) GPIO API.

rleh commented on src/modm/platform/spi/bitbang/bitbang_spi_master_impl.hpp:
```diff
@@ -53,9 +54,9 @@ modm::platform::BitBangSpiMaster<Sck, Mosi, Miso>::connect()
-	Sck::setOutput(Gpio::OutputType::PushPull);
-	Mosi::setOutput(Gpio::OutputType::PushPull);
-	Miso::setInput(Gpio::InputType::Floating);
+	Sck::setOutput(Sck::OutputType::PushPull);
+	Mosi::setOutput(Mosi::OutputType::PushPull);
+	Miso::setInput(Miso::InputType::Floating);
```
As far as I can see, that's not really necessary, is it?

henrikssn replied:
It is needed since `Gpio` is now a template, so it can't be used unspecialized like this anymore.

It makes sense as different `Gpio` can have different options for `InputType` / `OutputType` (e.g. not all pins supports pull-up/pull-down).

rleh commented on src/modm/platform/uart/sam/uart.hpp.in:
```diff
@@ -41,19 +41,31 @@ namespace platform
 class {{ name }} : public UartBase, public ::modm::Uart
 {
+
+	// Separate because that makes GCC print the template args.
```
Nice to know! :+1:

rleh commented on src/modm/platform/usb/sam/usb.hpp:
```diff
@@ -32,17 +32,20 @@ class Usb
-		Connector::connect();
+		using DpPin = GetPin_t<PeripheralPin::Dp, Pins...>;
+		using DmPin = GetPin_t<PeripheralPin::Dm, Pins...>;
+		static_assert(
+			!std::is_same_v<typename DpPin::Base, typename DmPin::Base>,
+			"Dp and Dm cannot use the same pin!");
```
Why not use the trick from above here too?
```suggestion
		template<class DpPin, class DmPin>
		struct ValidateNotSame {
			static_assert(
				!std::is_same_v<typename DpPin::Base, typename DmPin::Base>,
				"Dp and Dm cannot use the same pin!");
		};
```

henrikssn replied:
We require `DpPin` to use the `Dp` peripheral pin and `DmPin` to use `Dm`, so it can't happen that the pins use the same peripheral pin.

This is a problem on SAM UART because there the same signal (e.g. `Pad0`) can be used as both `Rx` and `Tx` peripheral pins.

rleh reviewed (changes requested):
Nice work! :+1: 

I have commented a few remarks and questions inline.

henrikssn reviewed (commented):
Thanks for the review!

rleh commented:
Merge @salkinium?

henrikssn committed:
[gpio] Refactor SAM signals API

henrikssn committed:
[tools] Add clang-format style

salkinium commented on tools/scripts/clang-format.sh:
```diff
@@ -0,0 +1,11 @@
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#----------------------------------------------------------------------------
+
+git diff --name-only --diff-filter=A -C -M develop | grep -e "\.\(c\|h\|hpp\|cpp\)\$" | xargs clang-format -i "$@"
```
Added a filter so that only newly added files get formatted, otherwise I go crazy as discussed ;-P

henrikssn replied:
Thanks! I realized this command is even better, as it reformats only changed lines: `git clang-format upstream/develop`

salkinium reviewed (approved):
Squashed and rebased, I'll merge once the CI passes.
